<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <!-- Turtlebot selection -->
    <arg name="model1" default="burger" doc="model type [burger, waffle, waffle_pi]"/>
    <arg name="model2" default="waffle" doc="model type [burger, waffle, waffle_pi]"/>
    
    <!-- Arg section -->
    <arg name="world"           default="w1"/>
    <arg name="first_tb3"       default="tb3_0"/>
    <arg name="second_tb3"      default="tb3_1"/>

    <arg name="first_tb3_x_pos" default="0.5"/>
    <arg name="first_tb3_y_pos" default="1.7"/>
    <arg name="first_tb3_z_pos" default="0.0"/>
    <arg name="first_tb3_yaw"   default="0.0"/>

    <arg name="second_tb3_x_pos" default="0.5"/>
    <arg name="second_tb3_y_pos" default="1.3"/>
    <arg name="second_tb3_z_pos" default="0.0"/>
    <arg name="second_tb3_yaw"   default="0.0"/>

    <!-- End Arg section -->

    <!-- Param section -->
    <rosparam file = "$(find exam)/config/param.yaml"/>

    <param name="topic_first_tb3"       value="$(arg first_tb3)"/>
    <param name="topic_second_tb3"      value="$(arg second_tb3)"/>
    <param name="first_tb3_x_pos"       value="$(arg first_tb3_x_pos)"/>
    <param name="first_tb3_y_pos"       value="$(arg first_tb3_y_pos)"/>

    <!-- End Param section -->

    <!-- Import of the simulation world -->
    <include file="$(find exploration_world)/launch/exp.launch">
        <arg name="world_id"    value="$(arg world)"/>
    </include>
    

    <!-- Turtlebot spawn -->
    <group ns = "$(arg first_tb3)">
        <param name="tf_prefix" value="$(arg first_tb3)_tf"/>
        <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model1).urdf.xacro" />
        <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" 
            args="-urdf -model $(arg first_tb3) 
            -x $(arg first_tb3_x_pos) 
            -y $(arg first_tb3_y_pos) 
            -z $(arg first_tb3_z_pos) 
            -Y $(arg first_tb3_yaw)
            -param robot_description" />
         
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
            <param name="publish_frequency" type="double" value="50.0" />
        </node>

    </group>

    <group ns = "$(arg second_tb3)">
        <param name="tf_prefix" value="$(arg second_tb3)_tf"/>
        <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model2).urdf.xacro" />
        <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" 
            args="-urdf -model $(arg second_tb3) 
            -x $(arg second_tb3_x_pos) 
            -y $(arg second_tb3_y_pos) 
            -z $(arg second_tb3_z_pos) 
            -Y $(arg second_tb3_yaw) 
            -param robot_description" />
        
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
            <param name="publish_frequency" type="double" value="50.0" />
        </node>

    </group>
    
    <!-- End Turtlebot spawn -->
    

    <!-- Teleop node 
    <node pkg="exam" type="key_teleop" name="key_teleop" output="screen"/> -->
    
</launch>
